rdm/0007-access-control.md

```diff
@@ -0,0 +1,222 @@
+
+Invenio RDM comes with a set of predefined generators:
+- Any user: Any user can perform the action.
+- Any user if public: Any user can perform the action if the record is public.
+- Record owners: Owners of the record can access perform the action.
+- Curators: Curators can perform an action.
+- Super Curators: ??
+
+
+**Permission Policies**
+
+A *permission policy* defines the set of generators upon which a certain action can be performed. A record policy would be set as:
+
+``` python
+RecordPermissionPolicy:
+    can_create: []
+    can_read: []
+    can_update: []
+    can_delete: []
+    can_list: []
+```
+
+The policy would return a set of *needs* and *excludes* for each action. The *needs* and *excludes* of a policy is the union of all those of the specified generators.
+
+For the following user story:
+
+```
+As an admin I want to grant read access to the owners of a document, except if the owner is part of team A.
+```
+
+The following policy applies the `RecordOwner` and `TeamMembership` generators to the record read action. The `RecordOwner` generator, allows the owners of the document (defined right above it, with ids 1, 2 and 3) to access the document. On the other hand, `TeamMembershipt` blacklists users belonging to Team A of accesing the document.
+
+``` python
+Document:
+    _owners: [1, 2, 3]
+
+RecordPermissionPolicy:
+    can_read: [RecordOwner, TeamMembership]
+
+RecordOwner:
+    needs:
+        return Document._owners
+
+TeanMembership:
+    needs:
+        pass
+    excludes:
+        return [Team A]
+```
+
+Assuming the following two users:
+
+``` python
+UserOne:
+    id: 1
+    team: B
+
+UserTwo:
+    id: 2
+    team: A
+```
+
+*UserOne* would be able to read the document since it has id `1`. However, *UserTwo* would not. Eventhough, it has id `2`, it belongs to `Team A` which has been blacklisted.
+
+## How we teach this
+
+This is a hard to teach concept because it involves a breaking change on the paradigm as from what it is today. However, it makes it much more user friendly and involves less work to implement custom access control rules. The best way to teach this would be to edit the corresponding section in the Invenio Tutorials (assuming the model is adopted by the whole Invenio Framework). In addition, create tutorial in the general RDM documentation.
+
+## Drawbacks
+
+It might increase the learning curve for current power users. However, for new adopters it will make the access control handling much easier.
+
+## Alternatives
+
+Keep using the current model where every instance has to implement their own using flask-principal.
+
+## Unresolved questions
+
+- Shall we implement *campus user* system role? It would require some sort of IP treatment.
+- Update and Delete actions over files, are iherited from the associated record. Or are per file?
```

Comment:
I have also had people ask about permission per file. A separate record could be made in this case with different permissions or all files should be given the same permission level as the most restricted file; i.e., per file seems excessive to me for now.